python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py

```diff
@@ -335,6 +385,62 @@ def _get_run_tags(self, run_id):
 
         return Tags(arn, cluster, cpu, memory)
 
+    def _run_task(self, run_task_kwargs, capacity_failure_strategy: dict, fallback_capacity_provider_strategy: list):
+
+        # initialize parameters
+        default_iteration = 1
+        retry_iterations = int(capacity_failure_strategy["maxRetries"]) if capacity_failure_strategy else 0
+        fallback_iteration = 1 if fallback_capacity_provider_strategy else 0
+        retry_interval = int(capacity_failure_strategy["retryInterval"]) if capacity_failure_strategy else 0
+
+        # fire off runs
+        total_iterations = default_iteration + retry_iterations + fallback_iteration
+        for i in range(total_iterations):
+
+            if fallback_capacity_provider_strategy and i == (total_iterations - 1):
+                run_task_kwargs["capacityProviderStrategy"] = fallback_capacity_provider_strategy
+
+            try:
+                response = self.ecs.run_task(**run_task_kwargs)
+                tasks = response["tasks"]
+                if not tasks:
+                    failures = response["failures"]
+                    if any(failure.get("reason") == "Capacity is unavailable at this time. Please try again later or in a different availability zone" for failure in failures):
+                        raise CapacityFailureException()
+                    else:
+                        exceptions = []
+                        for failure in failures:
+                            arn = failure.get("arn")
+                            reason = failure.get("reason")
+                            detail = failure.get("detail")
+                            exceptions.append(Exception(f"Task {arn} failed because {reason}: {detail}"))
+                        raise Exception(exceptions)
+
+                return tasks
+            
+            except CapacityFailureException:
```

Comment:
I don't think we should use an exception for expected business logic like this. Can we increment a counter directly instead?